### PR TITLE
Implement global table and table scan functionality

### PR DESCRIPTION
**Pull request by jadeallenx**
This PR implements both serial scans and parallel scans for Dynamo tables.

### CODE REVIEW

File: src/darcy.erl [hunk 1]
```diff
@@ -10,6 +10,9 @@
 -module(darcy).
 -include("darcy.hrl").
 
+-define(TIMEOUT, 5000).
+-define(BIG_TIMEOUT, 60000).
+
```
Comment (jonathn-al) [inline src/darcy.erl]: Maybe more descriptive names

File: src/darcy.erl [hunk 2]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
...
+    execute_scan(Client, make_scan_request([Expr, table_name(TableName)])).
+
+
+%% @doc This function executes a sequential scan over an entire table. In other
+%% words, it will continue to make new calls until no `LastEvaluatedKey' field
+%% is returned from Dynamo.
+%%
+%% The items will be accumulated into a list and returned.  This call is
+%% syntactic sugar for `scan/4' with a function of `same(X) -> X' passed
+%% in.
```
Comment (jonathn-al) [inline src/darcy.erl]: typo? syntactic sugar for `scan_all/4'

File: src/darcy.erl [hunk 3]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
 process_result_set(Other) ->
     {error, {query_error, Other}}.
 
+%% SCAN
+
+%% @doc This function executes a sequential table scan (no parallelism here)
+%% and returns results synchronously to the caller. If you want to scan a table
+%% with parallel workers, look at `scan/5'.
```
Comment (jonathn-al) [inline src/darcy.erl]: typo? look at `scan_parallel/5'

File: src/darcy.erl [hunk 4]
```diff
@@ -201,6 +211,57 @@ attempt_make_table(Client, Spec) ->
...
+make_global_table_if_not_exists(#{ region := Region } = Client,
+                                #{ <<"TableName">> := TableName } = Spec, Regions) ->
+    case lists:member(Region, Regions) of
+        false -> {error, {bad_region_spec, [Region, Regions]}};
+        true ->
+            ok = global_table_setup(Client, Spec, Regions),
+            attempt_make_global_table(Client, TableName, Regions)
+    end.
+
+global_table_setup(Client, Spec, Regions) ->
```
Comment (jonathn-al) [inline src/darcy.erl]: Not a problem, but would be neat if you included a more generic 'pmap' function (eg. similar to http://erlang.org/pipermail/erlang-questions/2009-January/041214.html) and used that here.

File: src/darcy.erl [hunk 5]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
...
+                  Fun  :: function() ) -> {ok, [ term() ]} |
+                                          {error, { Reason :: term(), Acc :: [ map() ]}}.
+scan_all(Client, TableName, Expr, Fun) ->
+    Request = make_scan_request([Expr, table_name(TableName)]),
+    do_scan_all(Client, Request, execute_scan(Client, Request), Fun, []).
+
+do_scan_all(_Client, _Req, {ok, empty_table}, _Fun, _Acc) -> {ok, []};
+do_scan_all(_Client, _Req, {ok, no_results}, _Fun, Acc) -> {ok, lists:reverse(lists:flatten(Acc))};
+do_scan_all(_Client, _Req, {ok, #{ <<"Items">> := I }}, Fun, Acc) ->
+    {ok, lists:reverse(lists:flatten([ lists:map(Fun, I) | Acc ]))};
```
Comment (jonathn-al) [inline src/darcy.erl]: Maybe I'm wrong, but should this be lists:flatten(lists:reverse(...)) instead?

File: src/darcy.erl [hunk 6]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
...
+do_scan_all(_Client, Req, {error, Error}, _Fun, Acc) ->
+    error_logger:error_msg("Error executing scan_all. request: ~p, error: ~p", [Req, Error]),
+    {error, {Error, Acc}};
+do_scan_all(Client, Req, {partial, #{ <<"LastEvaluatedKey">> := LEK, <<"Items">> := I }}, Fun, Acc) ->
+    NewRequest = make_scan_request([Req, #{ <<"ExclusiveStartKey">> => LEK }]),
+    do_scan_all(Client, NewRequest, execute_scan(Client, NewRequest), Fun, [ lists:map(Fun, I) | Acc]).
+
+%% @doc Scan a table in parallel using the given expression. This
+%% function is equivalent to `scan/7' with a timeout value of 60000
+%% milliseconds.
```
Comment (jonathn-al) [inline src/darcy.erl]: typo? equivalent to `scan_parallel/6'

File: src/darcy.erl [hunk 7]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
...
+            coordinator_loop( Pids -- [Pid], Reply, [ Partial | Acc ], [ Error | Errors ]);
+        Other ->
+            %% fall through for rando messages
+            error_logger:warning_msg("Received unexpected message: ~p", [Other]),
+            coordinator_loop(Pids, Reply, Acc, Errors)
+    after ?BIG_TIMEOUT * 5 ->
+        error_logger:info_msg("The scan has timed out."
+                              "Worker pids: ~p, reply pid: ~p, accumulator: ~p, errors: ~p",
+                              [Pids, Reply, Acc, Errors]),
+        Reply ! scan_timeout
```
Comment (jonathn-al) [inline src/darcy.erl]: should it be {error, scan_timeout} ?

File: src/darcy.erl [hunk 8]
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
...
+        Client :: darcy_request:aws_client(),
+        TableName :: binary(),
+        Expr :: map(),
+        Fun :: function(),
+        Timeout :: pos_integer(),
+        SegmentCount :: pos_integer(),
+        ReplyPid :: pid()) -> scan_timeout |
+                              { Results :: [ term() ], Errors :: [ term() ] }.
+scan_parallel(Client, TableName, Expr, Fun, _Timeout, SegmentCount, ReplyPid) ->
+    Reqs = [ make_scan_request([Expr, table_name(TableName), make_segments(N, SegmentCount)]) ||
```
Comment (jonathn-al) [inline src/darcy.erl:614]: Same comment as above, might be neat if you included a general 'parallel map with up to N workers' function and used that (?)

**Review by jonathn-al — APPROVED** [review]
Looking good

File: src/darcy.erl (hunk 1 above)
Comment (jadeallenx) [reply src/darcy.erl]: Do you have any good suggestions?

File: src/darcy.erl (hunk 2 above)
Comment (jadeallenx) [reply src/darcy.erl]: Thanks. Fixed.

File: src/darcy.erl (hunk 1 above)
Comment (jonathn-al) [reply src/darcy.erl]: how about TABLE_CREATION_TIMEOUT, TABLE_SCAN_TIMEOUT or similar?

File: src/darcy.erl (hunk 4 above)
Comment (jadeallenx) [reply src/darcy.erl]: Good idea.

File: src/darcy.erl (hunk 5 above)
Comment (jadeallenx) [reply src/darcy.erl]: Yeah, I made a function

File: src/darcy.erl (hunk 7 above)
Comment (jadeallenx) [reply src/darcy.erl]: hmm maybe. for now Im going to leave it as-is

**Comment by jadeallenx** [comment]
@jonathn-al Think I've implemented all of your suggestions.

**Comment by jonathn-al** [comment]
👍

**Pull request merged**